File: src/lib/recommendation.ts
/**
 * クラブ推薦エンジン
 *
 * 飛距離分布 x ハザードリスクの期待値計算に基づき、
 * 最適なクラブをランキング形式で推薦する。
 *
 * ADR-002: ルールベース (正規分布近似 + 閾値判定) を採用。
 */

import type { ClubType } from "../types/club";
import type { Hazard } from "../types/course";
import type { Coordinate, Shot } from "../types/shot";
import { euclideanDistanceYards } from "../types/shot";
import { mean, normalCdf, stdDev } from "./statistics";

/** ハザード種別ごとのペナルティコスト */
const HAZARD_PENALTY: Record<string, number> = {
	water: 2.0,
	bunker: 1.0,
	ob: 3.0,
	tree: 0.5,
	rough: 0.3,
};

/** コールドスタート用ジェネリック統計 (プロ平均ベース) */
const GENERIC_STATS: ReadonlyMap<ClubType, { mean: number; stdDev: number }> = new Map([
	["driver", { mean: 230, stdDev: 20 }],
	["3w", { mean: 210, stdDev: 18 }],
	["5w", { mean: 195, stdDev: 16 }],
	["7w", { mean: 180, stdDev: 15 }],
	["3i", { mean: 180, stdDev: 14 }],
	["4i", { mean: 170, stdDev: 13 }],
	["5i", { mean: 160, stdDev: 12 }],
	["6i", { mean: 150, stdDev: 11 }],
	["7i", { mean: 140, stdDev: 10 }],
	["8i", { mean: 130, stdDev: 9 }],
	["9i", { mean: 120, stdDev: 8 }],
	["pw", { mean: 110, stdDev: 7 }],
	["aw", { mean: 95, stdDev: 6 }],
	["sw", { mean: 80, stdDev: 5 }],
	["lw", { mean: 60, stdDev: 5 }],
	["putter", { mean: 10, stdDev: 3 }],
]);

/** 推薦結果 */
export interface ClubRecommendation {
	readonly club: ClubType;
	readonly expectedValue: number;
	readonly fairwayKeepRate: number;
	readonly hazardRisk: number;
	readonly distanceMean: number;
	readonly distanceStdDev: number;
	readonly isGeneric: boolean;
}

/** 推薦エンジンのインターフェース (将来の ML 移行用) */
export interface RecommendationEngine {
	recommend(
		currentPosition: Coordinate,
		targetPosition: Coordinate,
		hazards: readonly Hazard[],
		shots: readonly Shot[],
		roundCount: number,
	): readonly ClubRecommendation[];
}

/** クラブ別の飛距離統計を取得 */
function getClubDistanceStats(shots: readonly Shot[], club: ClubType): { mean: number; stdDev: number } | null {
	const distances = shots.filter((s) => s.club === club).map((s) => s.distanceYards);
	if (distances.length === 0) return null;
	return { mean: mean(distances), stdDev: stdDev(distances) };
}

/** フェアウェイ/グリーンオン確率を計算 */
function calculateFairwayRate(targetDistance: number, clubMean: number, clubStdDev: number, tolerance: number): number {
	if (clubStdDev <= 0) return targetDistance === clubMean ? 1 : 0;
	// ターゲット距離 +/- tolerance の範囲に入る確率
	const lower = normalCdf(targetDistance - tolerance, clubMean, clubStdDev);
	const upper = normalCdf(targetDistance + tolerance, clubMean, clubStdDev);
	return upper - lower;
}

/** ハザード到達確率を計算 (ユークリッド距離) */
function calculateHazardRisk(
	currentPosition: Coordinate,
	hazards: readonly Hazard[],
	clubMean: number,
	clubStdDev: number,
): number {
	if (hazards.length === 0) return 0;
	if (clubStdDev <= 0) return 0;

	let totalRisk = 0;
	for (const hazard of hazards) {
		const hazardDistance = euclideanDistanceYards(currentPosition, hazard.position);
		const hazardNear = hazardDistance - hazard.radiusYards;
		const hazardFar = hazardDistance + hazard.radiusYards;

		// ハザード範囲に飛距離が入る確率
		const probability = normalCdf(hazardFar, clubMean, clubStdDev) - normalCdf(hazardNear, clubMean, clubStdDev);
		const penalty = HAZARD_PENALTY[hazard.type] ?? 1.0;
		totalRisk += probability * penalty;
	}

	return totalRisk;
}

/** ルールベース推薦エンジン */
export class RuleBasedEngine implements RecommendationEngine {
	/** コールドスタート閾値 (ラウンド数) */
	private readonly coldStartThreshold: number;

	constructor(coldStartThreshold = 5) {
		this.coldStartThreshold = coldStartThreshold;
	}

	recommend(
		currentPosition: Coordinate,
		targetPosition: Coordinate,
		hazards: readonly Hazard[],
		shots: readonly Shot[],
		roundCount: number,
	): readonly ClubRecommendation[] {
		const targetDistance = euclideanDistanceYards(currentPosition, targetPosition);
		const isGeneric = roundCount < this.coldStartThreshold;
		const tolerance = 15; // ターゲットからの許容誤差 (ヤード)

		const recommendations: ClubRecommendation[] = [];

		for (const [club, genericStats] of GENERIC_STATS) {
			// putter はティーショット推薦対象外
			if (club === "putter" && targetDistance > 30) continue;

			let clubMean: number;
			let clubStdDev: number;
			let useGeneric: boolean;

			if (isGeneric) {
				clubMean = genericStats.mean;
				clubStdDev = genericStats.stdDev;
				useGeneric = true;
			} else {
				const userStats = getClubDistanceStats(shots, club);
				if (userStats && userStats.stdDev > 0) {
					clubMean = userStats.mean;
					clubStdDev = userStats.stdDev;
					useGeneric = false;
				} else {
					clubMean = genericStats.mean;
					clubStdDev = genericStats.stdDev;
					useGeneric = true;
				}
			}

			const fairwayKeepRate = calculateFairwayRate(targetDistance, clubMean, clubStdDev, tolerance);
			const hazardRisk = calculateHazardRisk(currentPosition, hazards, clubMean, clubStdDev);

			// 期待値 = フェアウェイキープ率 - ハザードリスク
			const expectedValue = fairwayKeepRate - hazardRisk;

			recommendations.push({
				club,
				expectedValue,
				fairwayKeepRate,
				hazardRisk,
				distanceMean: clubMean,
				distanceStdDev: clubStdDev,
				isGeneric: useGeneric,
			});
		}

		// 期待値降順でソート、上位3つを返す
		return [...recommendations].sort((a, b) => b.expectedValue - a.expectedValue).slice(0, 3);
	}
}

File: api/internal/service/recommendation.go
package service

import (
	"sort"

	"github.com/ryusei/fairway-caddie/api/internal/model"
)

// ハザード種別ごとのペナルティコスト。
var hazardPenalty = map[model.HazardType]float64{
	model.HazardWater:  2.0,
	model.HazardBunker: 1.0,
	model.HazardOB:     3.0,
	model.HazardTree:   0.5,
	model.HazardRough:  0.3,
}

// genericClubStats はコールドスタート用ジェネリック統計 (プロ平均ベース)。
type genericClubStats struct {
	Mean   float64
	StdDev float64
}

var genericStats = map[model.ClubType]genericClubStats{
	model.ClubDriver: {Mean: 230, StdDev: 20},
	model.Club3W:     {Mean: 210, StdDev: 18},
	model.Club5W:     {Mean: 195, StdDev: 16},
	model.Club7W:     {Mean: 180, StdDev: 15},
	model.Club3I:     {Mean: 180, StdDev: 14},
	model.Club4I:     {Mean: 170, StdDev: 13},
	model.Club5I:     {Mean: 160, StdDev: 12},
	model.Club6I:     {Mean: 150, StdDev: 11},
	model.Club7I:     {Mean: 140, StdDev: 10},
	model.Club8I:     {Mean: 130, StdDev: 9},
	model.Club9I:     {Mean: 120, StdDev: 8},
	model.ClubPW:     {Mean: 110, StdDev: 7},
	model.ClubAW:     {Mean: 95, StdDev: 6},
	model.ClubSW:     {Mean: 80, StdDev: 5},
	model.ClubLW:     {Mean: 60, StdDev: 5},
	model.ClubPutter: {Mean: 10, StdDev: 3},
}

// RecommendationService はクラブ推薦サービス。
type RecommendationService struct {
	coldStartThreshold int
}

// NewRecommendationService は新しい推薦サービスを作成する。
func NewRecommendationService(coldStartThreshold int) *RecommendationService {
	if coldStartThreshold <= 0 {
		coldStartThreshold = 5
	}
	return &RecommendationService{
		coldStartThreshold: coldStartThreshold,
	}
}

// Recommend はクラブ推薦を実行する。
func (rs *RecommendationService) Recommend(
	currentPosition model.Coordinate,
	targetPosition model.Coordinate,
	hazards []model.Hazard,
	shots []model.Shot,
	roundCount int,
) []model.ClubRecommendation {
	targetDistance := model.EuclideanDistanceYards(currentPosition, targetPosition)
	isGeneric := roundCount < rs.coldStartThreshold
	const tolerance = 15.0

	var recommendations []model.ClubRecommendation

	// 安定した順序のためAllClubTypesを使ってイテレートする
	for _, club := range model.AllClubTypes {
		gs, ok := genericStats[club]
		if !ok {
			continue
		}

		// putter はティーショット推薦対象外
		if club == model.ClubPutter && targetDistance > 30 {
			continue
		}

		var clubMean, clubStdDev float64
		var useGeneric bool

		if isGeneric {
			clubMean = gs.Mean
			clubStdDev = gs.StdDev
			useGeneric = true
		} else {
			userStats := getClubDistanceStats(shots, club)
			if userStats != nil && userStats.StdDev > 0 {
				clubMean = userStats.Mean
				clubStdDev = userStats.StdDev
				useGeneric = false
			} else {
				clubMean = gs.Mean
				clubStdDev = gs.StdDev
				useGeneric = true
			}
		}

		fairwayKeepRate := calculateFairwayRate(targetDistance, clubMean, clubStdDev, tolerance)
		hazardRisk := calculateHazardRisk(currentPosition, hazards, clubMean, clubStdDev)

		// 期待値 = フェアウェイキープ率 - ハザードリスク
		expectedValue := fairwayKeepRate - hazardRisk

		recommendations = append(recommendations, model.ClubRecommendation{
			Club:            club,
			ExpectedValue:   expectedValue,
			FairwayKeepRate: fairwayKeepRate,
			HazardRisk:      hazardRisk,
			DistanceMean:    clubMean,
			DistanceStdDev:  clubStdDev,
			IsGeneric:       useGeneric,
		})
	}

	// 期待値降順でソート
	sort.Slice(recommendations, func(i, j int) bool {
		return recommendations[i].ExpectedValue > recommendations[j].ExpectedValue
	})

	// 上位3つを返す
	if len(recommendations) > 3 {
		recommendations = recommendations[:3]
	}

	return recommendations
}

// getClubDistanceStats はクラブ別の飛距離統計を取得する。
func getClubDistanceStats(shots []model.Shot, club model.ClubType) *genericClubStats {
	var distances []float64
	for _, s := range shots {
		if s.Club == club {
			distances = append(distances, s.DistanceYards)
		}
	}
	if len(distances) == 0 {
		return nil
	}
	return &genericClubStats{
		Mean:   Mean(distances),
		StdDev: StdDev(distances),
	}
}

// calculateFairwayRate はフェアウェイ/グリーンオン確率を計算する。
func calculateFairwayRate(targetDistance, clubMean, clubStdDev, tolerance float64) float64 {
	if clubStdDev <= 0 {
		if targetDistance == clubMean {
			return 1
		}
		return 0
	}
	lower := NormalCdf(targetDistance-tolerance, clubMean, clubStdDev)
	upper := NormalCdf(targetDistance+tolerance, clubMean, clubStdDev)
	return upper - lower
}

// calculateHazardRisk はハザード到達確率を計算する。
func calculateHazardRisk(
	currentPosition model.Coordinate,
	hazards []model.Hazard,
	clubMean, clubStdDev float64,
) float64 {
	if len(hazards) == 0 || clubStdDev <= 0 {
		return 0
	}

	var totalRisk float64
	for _, hazard := range hazards {
		hazardDistance := model.EuclideanDistanceYards(currentPosition, hazard.Position)
		hazardNear := hazardDistance - hazard.RadiusYards
		hazardFar := hazardDistance + hazard.RadiusYards

		probability := NormalCdf(hazardFar, clubMean, clubStdDev) - NormalCdf(hazardNear, clubMean, clubStdDev)
		penalty, ok := hazardPenalty[hazard.Type]
		if !ok {
			penalty = 1.0
		}
		totalRisk += probability * penalty
	}

	return totalRisk
}
